theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import absolute_import, print_function, division
+
+import theano
+from theano import Op
+from theano import config
+import theano.tensor as T
+from .basic_ops import (gpu_contiguous, as_gpuarray_variable,
+                        infer_context_name, CGpuKernelBase)
+import theano.tensor.nnet.ctc
+from .type import GpuArrayType
+from .elemwise import GpuDimShuffle
+from theano.gradient import grad_undefined
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+
+import os
+import pygpu
+
+ctc_enabled = config.ctc.enabled
+
+
+class GpuConnectionistTemporalClassification(CGpuKernelBase, Op):
+    """
+    GPU wrapper for Baidu CTC loss function.
+    """
+    __props__ = ('context_name', 'compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_gpu)"
+
+    def __init__(self, compute_grad=True, context_name=None):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_gpu_no_grad)"
+        self.compute_grad = compute_grad
+        self.context_name = context_name
+
+        Op.__init__(self)
+        CGpuKernelBase.__init__(self, self.func_file, self.func_name)
+
+        self.costs = GpuArrayType(dtype='float32',
+                                  broadcastable=(False,),
+                                  context_name=self.context_name)
+
+        if self.compute_grad:
+            self.gradients = GpuArrayType(dtype='float32',
+                                          broadcastable=(False, False, False,),
+                                          context_name=self.context_name)
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the compiled library (libwarpctc.so) is available
+            # at the build directory of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "build"))
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray_helper.h', 'gpuarray/types.h',
+                'gpuarray_api.h', 'gpuarray/array.h', 'gpuarray/util.h']
+
```

Comment:
Now, to get the GPU context in C code (last parameter of functions, `PyGpuContextObject *  ctx`), we will set the Op param.
First we define the param as Op attribute, then we override the `get_params()` method that will retrieve the context from node inputs:
```python
# gpu_context_type must be imported from gpuarray/type.py
params_type = gpu_context_type
def get_params(self, node):
    # input 0 is `activations`, which will be on GPU, so it will have a context.
    return node.inputs[0].type.context
```